### DISCUSSION TITLE
Problems with flickering select-fields in Java application (Interactive Broker TWS)

### DISCUSSION THREAD

**Discussion by belugame** [i3/i3#5558] [Q&A]
Hey,

Is anyone here using Interactive Brokers' "IB Trader Workstation (TWS)" on i3wm? It is a java application that has worked without limitations for me for years. But since maybe half a year it has an annoying bug. I have reported it to IB but there was not real help from the support.

I'm on Debian trixie/sid with i3 4.22

Here is a screen cast of the problem: [https://www.dropbox.com/s/vrxb0mzos829wsn/simplescreenrecorder-2023-06-29_11.59.47.mp4?dl=0](https://www.dropbox.com/s/vrxb0mzos829wsn/simplescreenrecorder-2023-06-29_11.59.47.mp4?dl=0)
(Had to crop it for privacy of my account)

It happens when I click on a control element that opens a select field e.g. for the quantity or the limit price. These fields allow both selecting a value from the list or entering a value one self. The bug does not happen on fields that only allow selecing pre-given values. It most of the time shows the list for a fraction of a second and then it is gone again. That way it is impossible to select a value or enter one either. So each time I have to try up to two dozen times until it stays open.

I wonder if that is an incompatibility between i3wm and TWS or java applications since I have not found anyone else facing this issue. It is quite annoying, maybe some has a hint what could fix it? :)

@AntoniosHadji: I saw you are using TWS too, do you get that problem too?

**Comment by bsdice** [comment]
```
for_window [class="ibcalpha-ibc-IbcTws" title=".* Exit Session Setting"] floating enable border normal
for_window [class="ibcalpha-ibc-IbcTws" title="Order Entry"] floating enable border normal
for_window [class="ibcalpha-ibc-IbcTws" title="win*"] floating enable border normal
for_window [class="ibcalpha-ibc-IbcTws" title="Advanced Support"] floating enable border normal
no_focus [class="ibcalpha-ibc-IbcTws"]
```
Fellow TWS user... I am on Java 10.0.2 because of better HiDPI support than Java 8 and using my own updater/unpacker/installer to be fully portable on Linux, also using IBC helper. A partial solution is the no_focus line above, you won't be able to use the picker menu, but at least you can enter the values. HTH

The class above will be different when not using IBC, use https://gist.github.com/major/ff11ccdd73941109abb4ac2d35c976eb to figure it out.

Anything later than Java 10.0.2 is missing JavaFX and is also introducing new problems when going with Bellsoft full JDK. And for >8 you need various fixes because the security model was locked down more:

```
--illegal-access=warn \
--add-exports java.desktop/sun.awt=ALL-UNNAMED \
--add-exports javafx.graphics/com.sun.javafx.stage=ALL-UNNAMED \
--add-exports javafx.controls/com.sun.javafx.scene.control=ALL-UNNAMED \
--add-opens javafx.swing/javafx.embed.swing=ALL-UNNAMED \
--add-opens javafx.graphics/com.sun.javafx.stage=ALL-UNNAMED \
--add-opens java.base/java.lang=ALL-UNNAMED \
--add-opens java.base/java.util=ALL-UNNAMED \
--add-opens java.base/java.util.concurrent=ALL-UNNAMED \
--add-opens java.desktop/java.awt=ALL-UNNAMED \
--add-opens java.desktop/java.awt.dnd=ALL-UNNAMED \
--add-opens java.desktop/java.awt.event=ALL-UNNAMED \
--add-opens java.desktop/javax.swing=ALL-UNNAMED \
--add-opens java.desktop/javax.swing.event=ALL-UNNAMED \
--add-opens java.desktop/javax.swing.table=ALL-UNNAMED \
--add-opens java.desktop/javax.swing.plaf.basic=ALL-UNNAMED \
--add-opens java.desktop/javax.swing.colorchooser=ALL-UNNAMED \
--add-opens java.desktop/java.awt=ALL-UNNAMED \
--add-opens java.desktop/java.awt.peer=ALL-UNNAMED \
--add-opens java.desktop/java.awt.event=ALL-UNNAMED \
--add-opens javafx.graphics/com.sun.javafx.application=ALL-UNNAMED \
--add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED
```

**Reply by belugame** [reply]
Thank you very much for your input @bsdice 
I have since migrated to wayland & sway but have the very same issue there too. See: https://github.com/swaywm/sway/issues/7823

I will try out your suggestions on sway. Sorry, but what does "IBC" respectively IBC helper refer to?

**Comment by bsdice** [comment]
For IBC see here: https://github.com/IbcAlpha/IBC
It is automating logins except 2FA and automating some settings.

Since IBKR has put out the new IBKR Desktop application for Windows or Mac, I doubt an error such as this will get fixed for Linux. They couldn't be bothered to fix a simple CSS error when displaying the PAXOS icon that I reported. New charting based on embedded Chromium and custom TradingView is similarly a mess. But I have a good use for the IB API so I keep it.

Of note, I do no longer use the compositor picom because latest X11 beta has crisp and perfect double buffering against tearing in movies, and I have no use for swoosh swoosh eye strain on the desktop. But I have seen an effect when running it, that the readability of the window improves. Maybe some fade-in for the TWS window class can be concocted, that forces the dialog to top and not vanish after a few milliseconds.

I also tried all mentioned workarounds from https://wiki.archlinux.org/title/Java#Gray_window,_applications_not_resizing_with_WM,_menus_immediately_closing but to no avail.

**Comment by belugame** [comment]
Thank you again for your input. I hope they help someone else.
Sadly on sway the window rules don't seem to have an effect.
My hacky work around for entering orders is:

- click on the "advanced" button in the "Order Entry" window.
- click "Save" in the new popup
- double-click on the unsubmitted order in your Activity/Orders panel
- Now you get a different window called "Order Ticket" where changing the limit price and quantity works

**Comment by stanim** [comment]
I am having the same issue. Is ib-tws working in another window manager (eg gnome, kde, lxde, ...)?
